### PR TITLE
Modular Graph-Based Recipe Recommendation Pipeline

### LINKED ISSUE #4: Deep learning approach
Deep learning approach to recommend recipes

### PR DESCRIPTION
**Pull request by AkhilByteWrangler**
**To:** Meowth Team  
**From:** Akhil (@AkhilByteWrangler)
**Subject:** Addition of modular scripts for graph construction and GNN-based model training

---

## Summary

This pull request introduces a clean and modular structure for our deep learning recipe recommendation pipeline. The system is now organized into four Python scripts under `scripts/deep/`, each responsible for a distinct stage of the process — from preprocessing and graph construction to model definition and training.

---

## File Breakdown

- `build_graph.py`  
  Builds the heterogeneous graph using user interactions, recipe metadata, ingredients, and review text. It filters and samples the dataset, constructs the graph using `HeteroData`, and saves both the graph and corresponding user/recipe ID maps.

- `define_graph.py`  
  Contains helper functions for building graph edges, enriching interactions with sentiment, computing TF-IDF scores, and assigning node features. Used internally by `build_graph.py`.

- `pre_process_dl.py`  
  Handles preprocessing of raw interaction and recipe data. This includes parsing stringified lists and preparing the dataframes used in downstream scripts.

- `model.py`  
 This script defines the RecipeRecommenderGNN, a heterogeneous graph neural network built with PyTorch Geometric. It uses SAGEConv within HeteroConv layers to perform message passing across different node types. Recipes are represented with real-valued features, while other nodes (like users or ingredients) use learnable embeddings. The model learns to map users and recipes into a shared embedding space for recommendation.

 Training involves binary classification using positive interactions from the graph and randomly sampled negative pairs. The model is optimized using binary cross-entropy loss. After training, it's exported using TorchScript for efficient deployment.

 Evaluation is performed using RMSE on a test set of user–recipe pairs with ground truth ratings. Predictions are computed by scoring user and recipe embeddings, passed through a sigmoid and scaled to match the 0–5 rating range. The evaluation logic is included in the code, and the test data is expected to come from the original Kaggle dataset.

---

## Execution Order

To use the full pipeline:

1. **Run `build_graph.py`**  
   This prepares the filtered dataset and constructs the `HeteroData` object along with ID maps. The outputs are saved to disk.

2. **Run `model.py`**  
   This script loads the saved graph, trains the GNN on the user-recipe interactions, and exports the trained model in TorchScript format for deployment.

---

## Notes

- All scripts are GPU-compatible and will fallback to CPU if CUDA is not available.
- Output files (outputs/deep) include:
  - `recipe_gnn_script_100k.pt`
  - `recipe_graph_100k.pt`
  - `user_map.pkl`
  - `recipe_mapk.pkl`

---

Please review and test each step in sequence. Let me know if any integration points need adjustments.

Best,
Akhil (@AkhilByteWrangler)

### CODE REVIEW

File: scripts/deep/model.py [hunk 1]
```diff
@@ -1,60 +1,183 @@
...
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+data = data.to(device)
+
+# Ensure all expected edge types are defined in the graph
+for edge_type in data.metadata()[1]:
+    if edge_type not in data.edge_index_dict or data.edge_index_dict[edge_type] is None:
+        data.edge_index_dict[edge_type] = torch.empty((2, 0), dtype=torch.long, device=device)
+
+# === Initialize Model and Optimizer ===
+model = RecipeRecommenderGNN(metadata=data.metadata()).to(device)
```
Comment (haranku16) [inline scripts/deep/model.py]: Let's avoid dangling code by wrapping it up in a main function and invoking if `__name__ == '__main__'`

File: scripts/deep/model.py [hunk 2]
```diff
@@ -1,60 +1,183 @@
...
+
+# Build full input for export
+x_dict = {}
+for node_type in data.metadata()[0]:
+    if node_type == 'recipe':
+        x_dict[node_type] = data[node_type].x
+    else:
+        x_dict[node_type] = torch.arange(data.num_nodes_dict[node_type], device=device)
 
-    def predict(self, X: pd.DataFrame) -> np.ndarray:
```
Comment (haranku16) [inline scripts/deep/model.py:12]: The purpose of this abstraction is mainly for the evaluate script to polymorphically iterate over a list of models that support a common prediction/evaluation interface. Please wrap at least the output of your current modeling pipeline in a class that exposes the same evaluate function definition, and add to the evaluate script's list of models to generate an RMSE score.

File: scripts/deep/pre_process_dl.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import pandas as pd
+import ast  
+
+def load_datasets(inter_path, recipe_path, ingr_map_path):
```
Comment (haranku16) [inline scripts/deep/pre_process_dl.py:4]: Please add docstrings

File: scripts/deep/pre_process_dl.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    recipes['steps'] = recipes['steps'].apply(ast.literal_eval)
+
+    return interactions, recipes, ingr_map_df
+
+# load everything
+interactions, recipes, ingr_map = load_datasets(
+    'RAW_interactions.csv',
+    'RAW_recipes.csv',
+    'ingr_map.pkl'
+)
```
Comment (haranku16) [inline scripts/deep/pre_process_dl.py]: Let's wrap in a main function

File: scripts/deep/define_graph.py
Comment (haranku16) [inline scripts/deep/define_graph.py:1]: Let's wrap in a main function

File: scripts/deep/build_graph.py
Comment (haranku16) [inline scripts/deep/build_graph.py:1]: Let's wrap in a main function

File: scripts/deep/model.py [hunk 5]
```diff
@@ -1,60 +1,183 @@
...
+            conv = HeteroConv({
+                edge_type: SAGEConv((-1, -1), hidden_channels)
+                for edge_type in metadata[1]
+            }, aggr='sum')
+            self.convs.append(conv)
+
+        self.lin_user = Linear(hidden_channels, out_channels)
+        self.lin_recipe = Linear(hidden_channels, out_channels)
+
+    def forward(self, x_dict, edge_index_dict, num_nodes_dict=None):
```
Comment (haranku16) [inline scripts/deep/model.py:43]: Let's add docstrings where missing

**Review by haranku16 — CHANGES_REQUESTED** [review]
Thanks Akhil for your work here. It's a really interesting approach to combine elements of NLP with GNNs to better understand user-item relationships in the context of the text details of the items (e.g. ingredients, descriptions). I left a few comments mainly around style, maintaining the Model abstraction, and making sure that we run evaluation on this modeling approach to validate its performance compared to the naive approach.

File: scripts/deep/model.py (hunk 2 above)
Comment (AkhilByteWrangler) [reply scripts/deep/model.py:12]: Thanks, @haranku16 !

Wrapped the GNN pipeline in a RecipeGNNModel class that implements a shared evaluate() method, enabling polymorphic evaluation across models. This abstraction handles embedding generation and computes RMSE over user-recipe interactions.

We can now support a clean evaluate.py script that iterates over a list of models, each exposing the same evaluation interface. The GNN model has been added to that list, and it plugs right into the loop without special handling.

Let me know if we should extend this to support other metrics or evaluation sets.

File: scripts/deep/build_graph.py
Comment (AkhilByteWrangler) [reply scripts/deep/build_graph.py:1]: Thanks @haranku16 - this definitely helps make the codebase more modular and maintainable.
Things got a bit messy during the transition from notebook-style code to a proper script.

File: scripts/deep/define_graph.py
Comment (AkhilByteWrangler) [reply scripts/deep/define_graph.py:1]: Thanks @haranku16 — this definitely helps make the codebase more modular and maintainable.
Things got a bit messy during the transition from notebook-style code to a proper script,

File: scripts/deep/model.py (hunk 5 above)
Comment (AkhilByteWrangler) [reply scripts/deep/model.py:43]: Thanks @haranku16 — this definitely helps make the codebase more readable.

File: scripts/deep/model.py (hunk 1 above)
Comment (AkhilByteWrangler) [reply scripts/deep/model.py]: Thanks @haranku16 — this definitely helps make the codebase more modular and maintainable.
Things got a bit messy during the transition from notebook-style code to a proper script,

File: scripts/deep/pre_process_dl.py (hunk 3 above)
Comment (AkhilByteWrangler) [reply scripts/deep/pre_process_dl.py:4]: Thanks @haranku16 — this definitely helps make the codebase more readable.

File: scripts/deep/pre_process_dl.py (hunk 4 above)
Comment (AkhilByteWrangler) [reply scripts/deep/pre_process_dl.py]: Thanks @haranku16 — this definitely helps make the codebase more modular and maintainable.
Things got a bit messy during the transition from notebook-style code to a proper script.

File: scripts/deep/model.py [hunk 6]
```diff
@@ -98,86 +110,130 @@ def export_script_model(model, out_dict, path="recipe_gnn_script_100k.pt"):
...
+                x_dict[node_type] = torch.arange(self.data.num_nodes_dict[node_type], device=self.device)
 
-    # Binary cross-entropy loss on prediction
-    loss = F.binary_cross_entropy_with_logits(scores, labels)
-    loss.backward()
-    optimizer.step()
+        return self.model(x_dict, self.data.edge_index_dict, self.data.num_nodes_dict)
 
-    print(f"Epoch {epoch}/{EPOCHS} — Loss: {loss.item():.4f} — Time: {time.time() - start:.2f}s")
+    def evaluate(self):
```
Comment (haranku16) [inline scripts/deep/model.py]: Thanks for adding this evaluate function back, but can we refer to [these lines](https://github.com/DukeAIPI540Spring2025Meowth/recommender-demo/blob/main/scripts/util/model.py#L15-L38) and adapt accordingly? Also, could we add this model to the evaluate script and report the RMSE in the PR description?

Comment (AkhilByteWrangler) [reply scripts/deep/model.py]: Thanks for the keen eye @haranku16. Gotta get you on my startup team in the future! Resolved!

File: scripts/deep/model.py [hunk 7]
```diff
@@ -1,60 +1,264 @@
...
+            node_type: (
+                self.data[node_type].x if node_type == 'recipe'
+                else torch.arange(self.data.num_nodes_dict[node_type], device=self.device)
+            )
+            for node_type in self.data.metadata()[0]
+        }
+
+        return self.model(x_dict, self.data.edge_index_dict, self.data.num_nodes_dict)
+
+    def evaluate_from_index(self, test_df):
```
Comment (haranku16) [inline scripts/deep/model.py:230]: Where are you calling this function? Can you include the script to run this evaluation and also report the evaluation result in the PR description/conversation?

**Review by haranku16 — APPROVED** [review]
I can adapt your impl for the evaluation script as part of #5. Thanks for getting this approach working with such a high performance (0.53 RMSE)

**Comment by johnjoel2001** [comment]
Awesome work @AkhilByteWrangler!!

A few highlights I really liked: 
- The use of `HeteroData` and `HeteroConv` in model.py is well-structured and leverages PyG effectively for this recommendation use case.
- The preprocessing and string-parsing logic in `pre_process_dl.py is` neat and handles raw data well.

- The way you've enriched interactions with sentiment and used TF-IDF for node features in `define_graph.py `adds depth to the graph representation.

- I also appreciate the inclusion of `TorchScript` export for deployment and a clear evaluation strategy using RMSE with sigmoid-scaled predictions.

Everything looks good to me: well-organized, well-commented. Great job!
